pkg/threadas/threadas_test.go

```diff
@@ -0,0 +1,152 @@
+//go:build darwin
+// +build darwin
+
+package threadas
+
+import (
+	"runtime"
+	"syscall"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+const timeout = 100 * time.Millisecond
+
+func TestThreadAsNotRoot(t *testing.T) {
+	t.Parallel()
+
+	if syscall.Getuid() == 0 {
+		t.Skip("Skipping -- test requires not-root")
+	}
+
+	uids := getUidData()
+	fn := uids.GenerateTestFunc(t, "expected failure", assert.Equal)
+	require.Error(t, ThreadAs(fn, timeout, uint32(uids.Uid), uint32(uids.Gid)), "Fails when run as non-root")
+}
+
+func TestThreadAs(t *testing.T) {
+	t.Parallel()
+
+	if syscall.Getuid() != 0 {
+		t.Skip("Skipping -- test requires root")
+	}
+
+	targetUid := 501
+	targetGid := 20
+
+	targetUids := uidData{Uid: targetUid, Euid: targetUid, Gid: targetGid, Egid: targetGid}
+	fnTargetUidsEqual := targetUids.GenerateTestFunc(t, "matches target user", assert.Equal)
+	fnTargetUidsNotEqual := targetUids.GenerateTestFunc(t, "does not matches target user", assert.NotEqual)
+
+	myUids := getUidData()
+	fnMyUidsEqual := myUids.GenerateTestFunc(t, "matches my user", assert.Equal)
+	fnMyUidsNotEqual := myUids.GenerateTestFunc(t, "does not matches my user", assert.NotEqual)
+
+	// Be somewhat thoughtful in how parallel works here -- using
+	// t.Parallel will spawn a new thread, which potentially
+	// undermines some of what we're testing. But we do want the
+	// top level to be parallel, so that we have more thread
+	// churn.
+	for i := 1; i < 100; i++ {
+		t.Run("", func(t *testing.T) {
+			t.Parallel()
+
+			t.Run("baseline", func(t *testing.T) {
+				require.NoError(t, fnMyUidsEqual(), "no thread, matches my uid")
+				require.NoError(t, fnTargetUidsNotEqual(), "no thread, does not match target")
+			})
+
+			t.Run("change to target uid", func(t *testing.T) {
+				require.NoError(t, ThreadAs(fnTargetUidsEqual, timeout, uint32(targetUid), uint32(targetGid)), "match target")
+				require.NoError(t, ThreadAs(fnMyUidsNotEqual, timeout, uint32(targetUid), uint32(targetGid)), "not match mine")
+			})
+
+			t.Run("change to my uid", func(t *testing.T) {
+				require.NoError(t, ThreadAs(fnTargetUidsNotEqual, timeout, uint32(0), uint32(0)), "not match target")
+				require.NoError(t, ThreadAs(fnMyUidsEqual, timeout, uint32(0), uint32(0)), "matches mine")
+			})
+
+			t.Run("baseline after setuids", func(t *testing.T) {
+				require.NoError(t, fnMyUidsEqual(), "no thread, matches my uid")
+				require.NoError(t, fnTargetUidsNotEqual(), "no thread, does not match target")
+			})
+		})
+	}
+}
+
+func TestTimeout(t *testing.T) {
+	t.Parallel()
+
+	fn := func() error {
+		time.Sleep(10 * time.Second)
+		return nil
+	}
+
+	require.Error(t, ThreadAs(fn, timeout, KAUTH_UID_NONE, KAUTH_GID_NONE))
```

Comment:
if you want to make sure you're getting the timeout error as opposed to _any_ error, you can use https://pkg.go.dev/github.com/stretchr/testify/assert#EqualError

Reply:
I think you are using it correctly ... didn't think to upgrade testify to get access to `require.ErrorIs()` which is better than the `require.EqualError()` I suggested

Reply:
The part I'm unsure is how to do the comparison with the custom errors. Errors are pointers, and I don't really understand how go handles comparing them. Since baking in some context here (like, the seconds in the timeout), go consider `Timeout{t: 1}` different than `Timeout{t: 2}`. But I don't think they should be, so I defined the `Timeout.Is` to use a type assertion.

But this all feels like I'm doing something slightly wrong.